### ISSUE TITLE
How to pass args to slurm-submit.py? 

### ISSUE THREAD

**Issue by cmeesters** [Snakemake-Profiles/slurm#15]
Hi,

Basically, I would like to re-discuss issue #7, because: I am currently trying to establish snakemake on our cluster. My goal is to abstract as much as possible from users, such that they - eventually - are provided with a curated set of configurations and "just" have to write and run the snakefiles.

That would require to overwrite defaults on a per-user basis. At least the SLURM-account needs to be supplied somehow. Deducing from SLURM, as established in the script, is an idea. Unfortunately it only works with very regular users: One day gone by and retrieving the account with sacct is not longer possible.

I am also willing to back-contribute my work, once working, or any fixes I might have. Thank you for the great work you have already done.

So, there is one issue I am not able to figure out, so far: Can arguments be transferred from the snakefile to the slurm-submit(-advanced).py script ("resources" will not work for the account argument, as "account" is neither integer nor callable)? If so, how? 

Best,
Christian

**Comment by johanneskoester** [comment]
Hi Christian,
thanks for asking! 

> So, there is one issue I am not able to figure out, so far: Can arguments be transferred from the snakefile to the slurm-submit(-advanced).py script ("resources" will not work for the account argument, as "account" is neither integer nor callable)? If so, how?
params are accessible from the submit script. However, it would not be idiomatic to encode such information into the Snakefile, because it is platform specific, and the Snakefile is meant to be independent of the environment and the computational platform. However, I would say that the account can either be passed as environment variable, or read from a user specific config file. In both cases the submission script would take care of retrieving the value and the Snakefile itself does not need to be touched at all.

**Comment by percyfal** [comment]
Hi, the account can also be set in the cluster configuration file which is accessed by the advanced submission script:

```
__default__:
    account: myaccount
``` 

Hope that does what you need.

**Comment by cmeesters** [comment]
Hi,

thanks. Think I did not get my point across.

Putting the account-name into an environment variable is about the extend I expect from (my) users to do - yet, the issue of passing account names just serves as an illustration. Having it put into a configuration file would require one configuration per user or project group (it is not possible to infer a slurm account from the linux groups on our machine - although it would be nice if snakemake would be capable of accepting more than one config file and merge them. ;-) ). That would lead away from the idea of providing a (largely) curated system.

However, regardless of concepts I - personally - have in mind, some arguments should be "passable" somehow, resp. auto passed to the scheduler wrapper (here, slurm-submit.py). E.g. generic addons (that is any peculiar issues admins solve via spank plugins in slurm or exec scripts in LSF or ...) with generic flags cannot be address, right now or least not kept away from the user* (the same '--ramdisk' flag works different from our site to one 80 km away and others do not have this at all, yet implement the same feature with a different syntax). Hence my idea for a cluster specific configuration (template).

Parsing cluster settings as in the current version of slurm-submit-py (advanced or not) is insufficient in our case -  that can simply be extended by me and contributed back over time (this is why people at various site should try and give feedback, e.g. in terms of code - I do not expect a solution tailored for our cluster). E.g. "sinfo -e -O "memory,features_act" -h -p <partition>" gives truncated output on our machine.
Oh, and just the output argument: Without it the submit script assumes output_\\%j.log, which is not a directory and hence an error is produced when attempting to create it. Here, a rule may have an output dir, which is not auto-passed to the submit script (which would have to infer the dirname and do the test - but this is another diversion of mine, sorry).
 
As for the --ramdisk flag, I was thinking to implement a "--generic" flag or similar, which would not be the actual sbatch statement, but rather parsed and interpreted based on a separator token, to produce a variable number of sbatch flags (did I start on gres plugins already? or start of ad hoc filesystems? etc.). However, that leads back to the problem on how to supply args / parameters to the submit script or the configuration file.

Well, perhaps this is overly complicated. Contemplating about it gives me:
- the question: the idea to add the ability to merge multiple configs given on the command line - is it a good one, Johannes?
- (How) Can the parsing of the config file be amended to parse mostly arbitrary settings?

Does this make sense, somehow? Or do I miss the obvious?

Best,
Christian

* I teach a lot and my conclusion, so far, is that any data analyst brought up in a non-compute affine environment (regardless of life-science or other) has a hard time solving scientific questions; implementing complex hardware-related logic is putting off people from actually using HPC - and our team is pretty limited in resources, we cannot solve everything. Hence my desire to abstract as much as possible and gain

**Comment by mdehollander** [comment]
I have a related question. I thought to use the slurm profile in a global config at /etc/xdg. Then you can not hardcode the account name. Therefore I added a line to `slurm-submit.py` to get the username (https://github.com/mdehollander/slurm/blob/account/%7B%7Bcookiecutter.profile_name%7D%7D/slurm-submit.py#L120):
```
arg_dict["account"] = getpass.getuser()
```

For me this works. Would you also do something like this? Or is there a better way to use a global config and still retrieve the username/account name.

**Comment by cmeesters** [comment]
Thanks for the input. However, this is a non-generic solution: A SLURM-Account is neither a user nor inferable from a UNIX/LINUX account. That such strings are identical is coincidence. It works on your system and is hence a "work-on-my-system" solution, I am sorry to say.

Moreover, the remaining "sub-issues" can hardly be solved by extending add-on scripts, I think.

**Comment by mdehollander** [comment]
Ok, thanks. If there is a more general way to do this I am happy to hear it.

**Comment by johanneskoester** [comment]
My vision of profiles was that they should hide the entire platform-specific configuration from the user. In the sense that they are provided by the cluster admins ideally. From your question, it seems like you might want to provide different profiles to different user groups/projects. Technically, I would probably solve this cookiecutter. I.e., you have a profile template for your cluster, and upon deployment cookiecutter lets you choose and fill in the details, which are then composed automatically into the final profile on disk.

**Comment by johanneskoester** [comment]
Then, either users could answer the cookiecutter questions, or you would do it for them, depending on their needs and what you want to allow/provide to them.

**Comment by cmeesters** [comment]
Hi,

I am sorry for the late pick-up of this thread; I have been busy.

As for the place to ask questions: I am open to suggestions, yet the cookiecutter repo shows a (closed) issue, which refers to the profiles repo as the place to aks questions.

Yes, Johannes, 

> My vision of profiles was that they should hide the entire platform-specific configuration from the user. In the sense that they are provided by the cluster admins ideally.

this is my vision, too.

Now, in order to achieve this, there should be one (curated) configuration, for otherwise every user has to invoke the cookiecutter setup and start from scratch. Individual ammendments to this configuration (such as the account or test configuration parts for those not available inside the curated config) could be added with

`snakemake ... --config config1.json,config2.json` 

where the resulting dict could be updated from the second config file, ideally with a sanity check in between.

However, the need to pass information from snakemake to the submitting script - and that is a second part, potentially best discussed separately - arises from workflow parts better pooled into one: Imaging a step to map reads onto a reference. One could have a config entry to define partiton/queue, memory per node, etc. but not the number of nodes: if a wrapper does stage-in of references to avoid I/O contention (read: random-access on a global parallel file system) it could do that for all the read files separately (1 stage-in per job * N samples) or once (1 multi-node job with 1 stage-in*). Snakemake is aware of the number of samples in an input directory (or rather can be), this is knowledge one needs prior to submission.

One job per sample is not an option. One serial smp job for many samples neither.

Conceptually, I do not know how to achieve this or implement this into snakemake.

Best,
Christian



* I, for instance, maintain such a wrapper with auto-release of unused resources in SLURM. This is a) to fulfill a political requirement (rather multinode than many smp jobs) but also because b) it provides a better performance and c) is simpler to script with regard to further logic..

**Comment by cmeesters** [comment]
PS
The reason for asking here is in https://github.com/percyfal/slurm/issues/3
> I think it'd be better if you post issues on the SnakemakeProfiles page.

**Comment by johanneskoester** [comment]
@cmeesters the partitioning you describe sounds like it is a job for snakemake, not the profile. It seems to fit to an extension of DAG partitioning I have been planning for a while already. Basically, I would like to allow the following CLI call:

```
snakemake --partition star=5
```

which would tell Snakemake to put star jobs into batches of 5, and submit them together (in a job that has enough resources to run them in parallel (i.e. the sum of threads, sum of memory).

**Comment by cmeesters** [comment]
Thank you. Frankly, THIS would even be better than the idea I have / had, because the synchronization overhead would be lower in average.

Any timeline for this to be implemented?

**Comment by johanneskoester** [comment]
Hopefully, I will be able to work on it soon. But I can't give an estimate at the moment.

**Comment by mwort** [comment]
Hi @percyfal, thanks for this slurm profile, very helpful indeed.
Picking up on the discussion here again, with the snakemake `--cluster-config` option [deprecated now,](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html#cluster-configuration-deprecated), what would be the best way to parse per-rule/execution slurm arguments to the submit scripts in the future? As others, I was also desperately looking for ways to actually use the commandline interface built inside the scripts until I found that snakemake only ever parses the jobfile only (i.e. what is the point of the CLI in the submit scripts then?). Would be happy to implement and contribute if it helps.

**Comment by cmeesters** [comment]
Actually, together with Johannes, who is on parental leave now, I have been looking into this issue. He introduces loading environment modules, instead of conda packages, which is one of the prerequisites for cluster computing.

Also, I am implementing a thin wrapper to parse the configuration, provide a curated set of rules and workflows. I realize the need for a mostly generic implementation, but know, that some parts are left to cluster admins to work on. This gap can only be filled with additional docs. I would be happy to team up.

**Comment by cmeesters** [comment]
Actually, together with Johannes, who is on parental leave now, I have been looking into this issue. He introduces loading environment modules, instead of conda packages, which is one of the prerequisites for cluster computing.

Also, I am implementing a thin wrapper to parse the configuration, provide a curated set of rules and workflows. I realize the need for a mostly generic implementation, but know, that some parts are left to cluster admins to work on. This gap can only be filled with additional docs. I would be happy to team up.

**Comment by percyfal** [comment]
Hi @mwort and thanks, glad it is of use. You are quite right; there is no need for the CLI. The first implementation was based on the pbs-torque profile which also uses argparse; I never bothered to change since it worked and provided a way to map options to sbatch with defaults. However, given the many comments regarding the implementation it would probably be good to do a rewrite. In any case, sorry for the confusion.

@cmeesters are you indeed working on a rewrite of the profile? I just picked up on the environment modules feature, which will be hugely beneficial to work at our HPC.

**Comment by mwort** [comment]
Thanks for the background @percyfal. I would support a rewrite or rather a major refactoring (and merge) of the two submit scripts. I would really favour something as system agnostic as possible, i.e. mainly just a way to parse args from a snakemake cluster_config.yaml file to the actual sbatch call possibly with greater flexibility to set defaults in the cookiecutter template. I can push an example soon. @cmeesters if you have made a start already, why not push a WIP PR here?

**Comment by cmeesters** [comment]
@percyfal not quite, see #31 
@mwort because what is needed is a different approach, see #31

**Comment by percyfal** [comment]
Closing this issue due to PR #33.
